domain/blockdag/process.go

```diff
@@ -226,124 +228,225 @@ func (dag *BlockDAG) createBlockNodeFromBlock(block *util.Block) (
 
-	err := dag.checkBlockTransactionsFinalized(block, node, flags)
+	err := newNode.checkDAGRelations()
 	if err != nil {
 		return nil, err
 	}
 
-	if err = dag.checkFinalityViolation(node); err != nil {
+	err = dag.checkBlockTransactionsFinalized(block, newNode, flags)
+	if err != nil {
+		return nil, err
+	}
+
+	err = dag.checkBlockHasNoChainedTransactions(block, newNode, flags)
+	if err != nil {
 		return nil, err
 	}
 
 	if err := dag.validateGasLimit(block); err != nil {
 		return nil, err
 	}
 
-	newBlockPastUTXO, txsAcceptanceData, newBlockFeeData, newBlockMultiSet, err :=
-		node.verifyAndBuildUTXO(dag, block.Transactions(), isBehaviorFlagRaised(flags, BFFastAdd))
+	var isNewSelectedTip bool
+	if newNode.isGenesis() {
+		isNewSelectedTip = true
+	} else {
+		// If the new block is not the selected tip - it's not in virtual's selectedParentChain, therefore it's
+		// not going to be utxo-verified at this point
+		isNewSelectedTip = dag.selectedTip().less(newNode)
+		if !isNewSelectedTip {
+			dag.index.SetBlockNodeStatus(newNode, statusUTXONotVerified)
+		}
+	}
+
+	dbTx, err := dag.databaseContext.NewTx()
+	if err != nil {
+		return nil, err
+	}
+	defer dbTx.RollbackUnlessClosed()
+
+	if isNewSelectedTip {
+		err = dag.validateAndApplyUTXOSet(newNode, block, dbTx)
+		if err != nil {
+			if !errors.As(err, &(RuleError{})) {
+				return nil, err
+			}
+			dag.index.SetBlockNodeStatus(newNode, statusDisqualifiedFromChain)
+		} else {
+			isViolatingSubjectiveFinality, err := newNode.isViolatingFinality()
+			if err != nil {
+				return nil, err
+			}
+			if isViolatingSubjectiveFinality {
+				dag.index.SetBlockNodeStatus(newNode, statusUTXONotVerified)
+				dag.sendNotification(NTFinalityConflict, &FinalityConflictNotificationData{
+					ViolatingBlockHash: newNode.hash,
+					ConflictTime:       mstime.Now(),
+				})
+			}
+		}
+	}
+
+	err = dag.applyDAGChanges(newNode, selectedParentAnticone)
 	if err != nil {
-		return nil, errors.Wrapf(err, "error verifying UTXO for %s", node)
+		return nil, err
 	}
 
-	err = node.validateCoinbaseTransaction(dag, block, txsAcceptanceData)
+	chainUpdates, err := dag.updateVirtualAndTips(newNode, dbTx)
 	if err != nil {
 		return nil, err
 	}
 
-	virtualUTXODiff, chainUpdates, err :=
-		dag.applyDAGChanges(node, newBlockPastUTXO, newBlockMultiSet, selectedParentAnticone)
+	err = dag.saveChangesFromBlock(block, dbTx)
 	if err != nil {
-		// Since all validation logic has already ran, if applyDAGChanges errors out,
-		// this means we have a problem in the internal structure of the DAG - a problem which is
-		// irrecoverable, and it would be a bad idea to attempt adding any more blocks to the DAG.
-		// Therefore - in such cases we panic.
-		panic(err)
+		return nil, err
 	}
 
-	err = dag.saveChangesFromBlock(block, virtualUTXODiff, txsAcceptanceData, newBlockFeeData)
+	err = dbTx.Commit()
 	if err != nil {
 		return nil, err
 	}
 
+	dag.clearDirtyEntries()
```

Comment:
Because we can clearDirtyEntries only after we commited dbTx.
Now, dbTx is no longer the sole property of saveChangesFromBlock, so we can't commit dbTx inside it.